[CORE-307] Fix Spend Report Visibility for Billing Project Users (Workspace Owners)

kevinmarete opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/CORE-307

This PR ensures that Billing Project Users can view spend reports for workspaces they own on the billing project workspaces list page. Previously, these costs appeared as N/A, and users would receive a `403 Forbidden error` when attempting to access the billing project spend report. This was due to the requireProjectAction check for `readSpendReport`, which was restricted to project owners. I have updated this to `createWorkspace`, allowing workspace owners to access spend reports across all billing projects.

**Before**
<img width="1791" alt="Before" src="https://github.com/user-attachments/assets/f28afbb1-1149-48fb-8b72-1bb11c65b86b" />

**After**
<img width="1791" alt="After" src="https://github.com/user-attachments/assets/6392fa5b-fe5e-4c08-880c-afb436a84dec" />

---

**PR checklist**

- [X] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should publish a new official `rawls-model` and perform the corresponding dependency updates as specified in the [README](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model):
  - [ ] in the automation subdirectory
  - [ ] in workbench-libs
  - [ ] in firecloud-orchestration
- [x] Get two thumbsworth of PR review
- [X] Verify all tests go green, including CI tests
- [x] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

kevinmarete committed:
Replace read_spend_report with create_workspace action

kevinmarete committed:
Update tests

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -477,7 +477,9 @@ class SpendReportingService(
     validateReportParameters(start, end)
-    requireProjectAction(project, SamBillingProjectActions.readSpendReport) {
+    requireProjectAction(project,
+                         SamBillingProjectActions.createWorkspace
```
`readSpendReport` sounds like the right action to be checking. This probably should be a sam config change to grant `readSpendReport` to the right workspace role.

dvoet replied:
reading the jira more closely, it sounds like we have some inconsistent permission checking. Here, we are checking a project action but in the consolidated spend report it is perhaps checking a workspace action? We should make the permission checks consistent

kevinmarete replied:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/bd3cd9f1f92da9525e4e262a0778a4b893fb73a2

kevinmarete committed:
Show billing project spend report to workspace owners

kevinmarete committed:
Merge branch 'develop' into km_CORE-307_update_billing_project_spend_report

kevinmarete committed:
Fix SpendReportingServiceSpec Tests

kevinmarete committed:
Format Test File

kevinmarete committed:
Applied correct formatting

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -470,32 +469,68 @@ class SpendReportingService(
-        projectNames <- getWorkspaceGoogleProjects(project)
+    for {
+      spendExportConf <- getSpendExportConfiguration(project)
+      projectNames <- getOwnedWorkspaceGoogleProjectsInProject(project, childContext)
```
`getWorkspaceGoogleProjects` is now unused except in a test - clean it up?

kevinmarete replied:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/802413ba4a91edce2ca73e664deee052fad7bd79

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -470,32 +469,68 @@ class SpendReportingService(
+    projectName: RawlsBillingProjectName,
+    childContext: RawlsRequestContext
+  ): Future[Map[GoogleProjectId, WorkspaceName]] =
+    requireProjectAction(projectName, SamBillingProjectActions.createWorkspace) {
```
this is still checking `create_workspace` on the project instead of [read_spend_report](https://github.com/broadinstitute/sam/blob/d88b19fae8fdc160d1f42120bce61de85dcaef2d/src/main/resources/reference.conf#L756-L758); this mixes up two separate concepts (creating a workspace vs. reading a spend report)

1. Do we need a permission check on the billing project at all? The code is already checking permissions on each workspace in the project.
2. If we do need to check permissions on the project itself, could we modify Sam's config to give the `read_spend_report` action to the appropriate users, and check for that action?

dvoet replied:
I think this will make more sense if we change our thinking from "show spend report for all workspaces in a billing project" to "show spend report for all workspaces _filtered by_ billing project". If you do that, then you see that the `getOwnedWorkspaceGoogleProjects` function above does everything you need since it is already grouped by billing project.

kevinmarete replied:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/290b38d49a0b1c46cb4d53ee8660ad94487bf369

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -470,32 +469,68 @@ class SpendReportingService(
     bytesProcessedCounter += stats.getEstimatedBytesProcessed
   }
 
+  def getOwnedWorkspaceGoogleProjects(
```
the naming is off here, these are not owned workspaces, they are workspaces where the caller can `readSpendReport`

kevinmarete replied:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/290b38d49a0b1c46cb4d53ee8660ad94487bf369

kevinmarete committed:
Cleanup getWorkspaceGoogleProjects

kevinmarete committed:
Refactor and use readSpendReport

kevinmarete committed:
Merge branch 'develop' into km_CORE-307_update_billing_project_spend_report

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -470,32 +450,85 @@ class SpendReportingService(
+          workspaceServiceConstructor(childContext).getGCPWorkspacesByBillingProjects(validWorkspaceIds)
+      }
+
+  def getSpendReportableWorkspaceGoogleProjectsInBillingProject(
+    projectName: RawlsBillingProjectName,
+    childContext: RawlsRequestContext
+  ): Future[Map[GoogleProjectId, WorkspaceName]] = {
+
+    def filterProjects(projects: Map[RawlsBillingProjectName, Seq[Workspace]]): Map[GoogleProjectId, WorkspaceName] =
+      projects.flatMap { case (_, value) =>
+        value.collect {
+          case workspace if workspace.namespace == projectName.value =>
+            workspace.googleProjectId -> workspace.toWorkspaceName
+        }
+      }
+
+    for {
+      projects <- getSpendReportableWorkspaceGoogleProjects(childContext)
+      filteredProjects = filterProjects(projects)
+      hasAction <- samDAO.userHasAction(SamResourceTypeNames.billingProject,
+                                        projectName.value,
+                                        SamBillingProjectActions.readSpendReport,
+                                        childContext
+      )
+      result <-
+        if (hasAction || filteredProjects.nonEmpty) {
+          Future.successful(filteredProjects)
+        } else {
+          Future.failed(
+            RawlsExceptionWithErrorReport(
+              StatusCodes.Forbidden,
+              s"${childContext.userInfo.userEmail.value} cannot perform ${SamBillingProjectActions.readSpendReport.value} on project ${projectName.value}"
+            )
+          )
+        }
+    } yield result
+  }
```
This is far too complicated. Checking the billing project action is not required, access at the workspace level has already been checked and that is sufficient. The filter function does not need to cycle through all the workspaces because the map is already keyed by billing project.

This function should be 
```
getSpendReportableWorkspaceGoogleProjects(childContext).map { workspacesByProject =>
  workspacesByProject.get(projectName).getOrElse(Map.empty).map { workspace =>
    workspace.googleProjectId -> workspace.toWorkspaceName
  }.toMap
}
```

kevinmarete replied:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/6ca541dbe6e14fd161620ceb45f298cda6aabc57

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -470,32 +450,85 @@ class SpendReportingService(
+        childContext
+      )
+      .flatMap {
+        case ownerWorkspaces if ownerWorkspaces.isEmpty =>
+          Future.successful(Map.empty[RawlsBillingProjectName, Seq[Workspace]])
```
Why is this case necessary? Will `getGCPWorkspacesByBillingProjects` fail if `validWorkspaceIds` is empty?

kevinmarete replied:
Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/6ca541dbe6e14fd161620ceb45f298cda6aabc57

kevinmarete committed:
Remove unnecessary isEmpty case

kevinmarete committed:
Refactored project filter

davidangb reviewed (approved):
Thanks for seeing this one through!

I like the unification of how the per-billing-project spend report and the consolidated spend report identify their workspaces; nice job on that.

Do all the deleted unit tests deserve deletion? Are any of those still valid?

kevinmarete committed:
Reviewed tests

kevinmarete commented:
> 98345d0



> Thanks for seeing this one through!
> 
> I like the unification of how the per-billing-project spend report and the consolidated spend report identify their workspaces; nice job on that.
> 
> Do all the deleted unit tests deserve deletion? Are any of those still valid?

Fixed: https://github.com/broadinstitute/rawls/pull/3193/commits/98345d04a44bee6da37f9cba9312abad8814968d I reviewed the five tests I had removed. I fixed and restored three of them, added one new test, and confirmed that the remaining two are no longer valid cases.
